### ISSUE TITLE
Make a better plan for relative paths in example-datasets

### ISSUE THREAD

**Issue by ewallace** [riboviz/example-datasets#27]
The way that relative paths are encoded in our `config.yaml` is needlessly confusing and needs addressing.

E.G. currently the file for Brar-2012,
`fungi/saccharomyces/Brar_2012_Meiosis_RPF_6-samples_CDS_w_250utrs_config.yaml`, gives the path to annotation files as:
`../../riboviz/example-datasets/fungi/saccharomyces/annotation/Saccharomyces_cerevisiae_yeast_CDS_w_250utrs.fa`
even though from where the yaml file sits, that file is _relatively_ in `annotation/Saccharomyces_cerevisiae_yeast_CDS_w_250utrs.fa`.

So, the config.yaml file assumes that the script is being called from another directory with a particular relationship to the location of `example-datasets`. That is not good.

We need a lightweight solution that doesn't need lots of relative-paths in the job submission script. That means, a documented plan for the relative locations of:

* example-datasets
* input/output/etc
* the riboviz executable

Would it work if we symlinked to the example-datasets repo from the directory that invokes? Or to the subfolder with the relevant dataset? Or what?

We should discuss in concert with [riboviz#228](/riboviz/riboviz/issues/228). This is sort-of a repeat of #13, but with more information.

**Comment by kavousan** [comment]
What about defining an "environment" variable and using that to shorten paths, is it supported?

**Comment by mikej888** [comment]
The following is the current state (with examples from the vignette and example data configurations):

The following files can be located anywhere:

```
asite_disp_length_file: data/yeast_standard_asite_disp_length.txt
codon_positions_file: data/yeast_codon_pos_i200.RData
features_file: data/yeast_features.tsv
orf_fasta_file: vignette/input/yeast_YAL_CDS_w_250utrs.fa
orf_gff_file: vignette/input/yeast_YAL_CDS_w_250utrs.gff3
rrna_fasta_file: vignette/input/yeast_rRNA_R64-1-1.fa
t_rna_file: data/yeast_tRNAs.tsv
```

The paths specified can be absolute or relative to the current directory in which RiboViz is invoked.

There is no assumption that these files are indeed in `data/` or `vignette/input/` or are co-located (the 7 files above could be in 7 different directories).

The following directories can be located anywhere and the path specified in the YAML can be absolute, or relative to the current directory:

```
dir_index: vignette/index # Built indices directory
dir_in: vignette/input # Input directory
dir_out: vignette/output # Output directory
dir_tmp: vignette/tmp # Intermediate files directory
```

The following files are assumed to be within dir_in:

```
fq_files:
  WTnone: SRR1042855_s1mi.fastq.gz
  WT3AT: SRR1042864_s1mi.fastq.gz
multiplex_fq_files:
- multiplex_umi_barcode_adaptor.fastq
sample_sheet: multiplex_barcodes.tsv
```

The following are now misnomers:

```
orf_index_prefix: YAL_CDS_w_250
rrna_index_prefix: yeast_rRNA
```

These are used as index file-name prefixes not paths.

The following can be ignored, as these relate to the soon-to-be-removed Python workflow:

```
cmd_file: run_riboviz_vignette.sh
dir_logs: vignette/logs
```

**Comment by mikej888** [comment]
I'd agree about removing any relative dependencies from the YAML files. All files in the YAML can refer to files or symbolic links that are within the same directory as the YAML file (or subdirectories). The YAML doesn't care if it's an actual file or a symbolic link.

On a related note, Nextflow does not need to be run within the "riboviz" directory. I'd documented [Invoking the workflow from outwith the RiboViz home directory](https://github.com/riboviz/riboviz/blob/develop/docs/user/prep-riboviz-run-nextflow.md#invoking-the-workflow-from-outwith-the-riboviz-home-directory).

I've now updated the example to use symbolic links as suggested above.

**Comment by ewallace** [comment]
I have a suggestion:

* Everything in `example-datasets` that is referred to by a config file (annotations, barcodes, etc) should be in the directory of the `config.yaml`, or a subdirectory.
* Can we recommend symlinking the relevant directory/files/subdirectories prior to a run?
* Alternatively, could we add the directory containing `config.yaml` to PATH prior to a run? That might be easier.

It's just getting the details right.

**Comment by mikej888** [comment]
> Could we add the directory containing config.yaml to PATH prior to a run? (I think that's what @kavousan meant by using an environment variable to shorten paths)

The Nextflow workflow (like the Python workflow) will consider any relative paths in the YAML as relative to the directory in which Nextflow (or Python) is invoked. For Nextflow this does not have to be in the riboviz directory. If Nextflow (or Python) is invoked in the directory which contains config.yaml then all is fine. This is the example given at https://github.com/riboviz/riboviz/blob/develop/docs/user/prep-riboviz-run-nextflow.md#invoking-the-workflow-from-outwith-the-riboviz-home-directory.

It's unclear to me how adding the directory containing config.yaml to the PATH helps.

Is the goal to allow the user to run Nextflow from outwith the directory containing config.yaml? For example:

```console
$ nextflow run <RIBOVIZ_DIR>/prep_riboviz.nf -params-file <SOME_DIR>/config.yaml -ansi-log false
```

**Comment by ewallace** [comment]
**Yes**, my goal is to be able to run Nextflow from outwith the directory containing `config.yaml`, **without** copying the yaml and annotation files from `example-datasets` to somewhere else, and with "as simple as possible" directory names.

Would symlinking the config.yaml directory work?

**Comment by mikej888** [comment]
It could work, it'd require trying it to see what happens.

**Comment by ewallace** [comment]
Created branch `annotation-paths-27` to test this. Will start with `simulated/mok/Mok-simYAL5_config.yaml`, then afterwards try some yeast data. Will require updates to Eddie documentation in the riboviz repository, to explain the relevant symlinking/setup procedure.

**Comment by ewallace** [comment]
I would like to discuss this at next dev meeting. This is important to solve, because everyone using riboviz currently has to spend too much time getting files in the right place, and we want to have a simple policy that saves this angst. And there are too many options.

Options include:
* Adding `example-datasets` and `riboviz` folders to [nextflow scope environment](https://www.nextflow.io/docs/latest/config.html#scope-env), supplied in a generic nextflow configuration file as in https://github.com/riboviz/riboviz/issues/232. Then, give paths relative to `example-datasets` root directory. E.G. `fungi/saccharomyces/annotation/Saccharomyces_cerevisiae_yeast_CDS_w_250utrs.fa`
* Insist that users symlink everything they need into `dir_in`. But then how do we get the `dir_in` on the search path?
* symlinking the config.yaml directory as discussed above.

Questions:
* What is really going to cause the fewest problems?
* Are we using `here::here` for the R code, and do we need to play nice with that? Or are we only passing fully resolved filenames into R scripts from the command line?

**Comment by mikej888** [comment]
I thought symlinking was the solution? I thought this was solved. Relative paths with ".." in the YAML are problematic as they are too brittle. Relative paths from the YAML to the current directory, where the current directory has the required files, or symlinks to these would seem to me to be the most usable.

Edward: "This issue isn't solved because I didn't implement the symlinking solution. The problem with symlinking is that AFAIK you can't easily symlink a file and a subdirectory in a single command. And we have many component files to input in total. So I thought it might be easier just to tell nextflow to look for the example-datasets and riboviz repositories, then to search within them. Then the same solution works for run after run."

Using the "scope" environment might be promising. But then every user would have to edit nextflow.config to set this up with environment variables specific to their own environment.

Edward: Yes - is that a problem though?

Bit of a pain for the user , one extra file to edit. I just tried something else. Within Nextflow:

```
example="${EXAMPLE}"
print(example)
```
Running
```console
$ EXAMPLE=123 nextflow run prep_riboviz.nf -params-file vignette/vignette_config.yaml -ansi-log false --publish_index_tmp --validate_only
```
prints
```
123
```
A challenge is how the environment variable (whether passed as above or via a Nextflow 'scope' environment) is then associated with the paths in dir_in/ and input/ in the YAML.

**Comment by mikej888** [comment]
Edward: Indeed, yes, that is a challenge it would be better to avoid.
OK I tried something else...same Nextflow invocation as above plus, in vignette_config.yaml:
```
example_relative_path: ${EXAMPLE}/example.txt
```
And in prep_riboviz.nf:
```
print("=====\n")
example="${EXAMPLE}"
print(example)
print("\n")
print(params.example_relative_path)
print("\n")
new_path = params.example_relative_path.replace('${EXAMPLE}', example)
print(new_path)
print("\n")
print("=====\n")
```
This gives:
```
=====
123
${EXAMPLE}/example.txt
123/example.txt
=====
i.e. the token '${EXAMPLE}' is replaced by the value of that environment variable.
So could we have users define RIBOVIZ_INPUT and RIBOVIZ_DATA and define variables to replace those tokens in config.yaml paths thereby inserting the paths into the configuration.

**Comment by mikej888** [comment]
This would have implications for using the regression test framework as it too would have to be updated to do similar replacements.

**Comment by ewallace** [comment]
That would work. Comparing solutions here; feel free to edit this post.

### Symlinking to relevant files and directories

pro: 
* very specific
* no copying of files
* flexible to users' directory structure

con: 
* requires setting up new symlinks (several lines of code) for every run for every dataset

### Specifying repos in Nextflow config file

pro: 
* no copying of files
* nfconfig files only need to be edited once per user
* can give consistent paths in all config.yamls in `example-datasets`

con: 
* specification is "outside" riboviz repos and workflow call
* still presumes a certain directory structure
* could get confused between `riboviz` and `riboviz/riboviz` if we also want to specify the code directory

### Define tokens in `config.yaml`, to be replaced on command line

pro: 
* it works
* flexible and specific
* no copying of files

con: 
* update regression test
* looks like programming inside `config.yaml` instead of a flat text file
* requires path specification in every nextflow function call

**Comment by mikej888** [comment]
There are three classes of data so three environment variables may be e.g. see https://github.com/riboviz/example-datasets/blob/weinberg_2016_dataset-20/fungi/saccharomyces/Gupta_2018_tRNA_Modification_Carbon_Nitrogen_Metabolism_RPF_9-samples_CDS_w_250utrs_config.yaml:

```
dir_in: G-Sc_2018/input

asite_disp_length_file: data/yeast_standard_asite_disp_length.txt
features_file: data/yeast_features.tsv

orf_fasta_file: ../../riboviz/example-datasets/fungi/saccharomyces/annotation/Saccharomyces_cerevisiae_yeast_CDS_w_250utrs.fa datasets/fungi/saccharomyces/annotation/Saccharomyces_cerevisiae_yeast_CDS_w_250utrs.gff3
rrna_fasta_file: ../../riboviz/example-datasets/fungi/saccharomyces/contaminants/Saccharomyces_cerevisiae_yeast_rRNA_R64-1-1.fa
sample_sheet: ../../riboviz/example-datasets/fungi/saccharomyces/annotation/barcodes/Gupta_2018_tRNA_Modification_Carbon_Nitrogen_Metabolism_RPF_9-samples_CDS_w_250utrs_config_barcode.tsv
```
example-datasets could be updated to duplicate datasets from the riboviz repository so that example-datasets has no dependence on any data in the riboviz repository.

How many environment variables are required and how are these named?

**Comment by ewallace** [comment]
I think this was solved by #274?

**Comment by kavousan** [comment]
Agreed to close.
